packages/preferences/src/browser/preferences-tree-widget.ts

```diff
@@ -236,36 +288,113 @@ export class PreferencesEditorsContainer extends DockPanel {
+                this._workspacePreferenceEditorWidget.close();
+                this._workspacePreferenceEditorWidget.dispose();
+            }
+            this._workspacePreferenceEditorWidget = newWorkspacePreferenceEditorWidget;
+        }
+    }
 
+    protected async getWorkspacePreferenceEditorWidget(): Promise<PreferencesEditorWidget | undefined> {
         const workspacePreferenceUri = await this.workspacePreferenceProvider.getUri();
         const workspacePreferences = workspacePreferenceUri && await this.editorManager.getOrCreateByUri(workspacePreferenceUri) as PreferencesEditorWidget;
 
         if (workspacePreferences) {
             workspacePreferences.title.label = 'Workspace Preferences';
             workspacePreferences.title.caption = await this.getPreferenceEditorCaption(workspacePreferenceUri!);
+            workspacePreferences.title.icon = 'database-icon medium-yellow file-icon';
+            workspacePreferences.editor.setLanguage(JSONC_LANGUAGE_ID);
             workspacePreferences.scope = PreferenceScope.Workspace;
-            this.addWidget(workspacePreferences);
         }
+        return workspacePreferences;
+    }
 
-        this.activatePreferenceEditor(this.preferenceScope);
-        super.onAfterAttach(msg);
-        this.onInitEmitter.fire(undefined);
+    get activeFolder(): string | undefined {
+        if (this._foldersPreferenceEditorWidget) {
+            return this._foldersPreferenceEditorWidget.editor.uri.parent.parent.toString();
+        }
+    }
+
+    async refreshFoldersPreferencesEditorWidget(currentFolder: string | undefined): Promise<void> {
+        const folders = this.workspaceService.tryGetRoots().map(r => r.uri);
+        const newFoldersPreferenceEditorWidget = await this.getFoldersPreferencesEditor(currentFolder || folders[0]);
+        if (newFoldersPreferenceEditorWidget) {
+            this.addWidget(newFoldersPreferenceEditorWidget,
+                { ref: this._workspacePreferenceEditorWidget || this._workspacePreferenceEditorWidget || this._userPreferenceEditorWidget });
+            if (this._foldersPreferenceEditorWidget) {
+                this._foldersPreferenceEditorWidget.close();
+                this._foldersPreferenceEditorWidget.dispose();
+            }
+            this._foldersPreferenceEditorWidget = newFoldersPreferenceEditorWidget;
+            this.onFolderPreferenceEditorUriChangedEmitter.fire(newFoldersPreferenceEditorWidget.editor.uri.toString());
+        }
+    }
+
+    protected async getFoldersPreferencesEditor(folder: string | undefined): Promise<PreferencesEditorWidget | undefined> {
+        if (this.workspaceService.saved) {
+            const settingsUri = await this.getFolderSettingsUri(folder);
+            const foldersPreferences = settingsUri && await this.editorManager.getOrCreateByUri(settingsUri) as PreferencesEditorWidget;
+            if (foldersPreferences) {
+                foldersPreferences.title.label = 'Folders Preferences';
```

Comment:
name too long, just use the first part: i.e., User, Workspace, Folders

Reply:
as discussed i will shorten texts in tab headers, and move "xxxx Preferences" into the tooltips.